src/Middleware/HttpLogging/src/MediaTypeOptions.cs

```diff
@@ -0,0 +1,127 @@
+            options.AddText("application/xml", Encoding.UTF8);
+            options.AddText("application/*+xml", Encoding.UTF8);
+            options.AddText("text/*", Encoding.UTF8);
+
+            return options;
+        }
+
+        /// <summary>
+        /// Adds a <see cref="MediaTypeHeaderValue"/> to be used for logging as text.
+        /// </summary>
+        /// <param name="mediaType">The MediaType to add.</param>
+        public void AddText(MediaTypeHeaderValue mediaType)
+        {
+            if (mediaType == null)
+            {
+                throw new ArgumentNullException(nameof(mediaType));
+            }
+
+            mediaType.Encoding ??= Encoding.UTF8;
+
+            _mediaTypeStates.Add(new MediaTypeState(mediaType));
+        }
+
+        /// <summary>
+        /// Adds a contentType to be used for logging as text.
+        /// </summary>
+        /// <remarks>
+        /// If charset is not specified in the contentType, the encoding will default to UTF-8.
+        /// </remarks>
+        /// <param name="contentType">The content type to add.</param>
+        public void AddText(string contentType)
+        {
+            if (contentType == null)
+            {
+                throw new ArgumentNullException(nameof(contentType));
+            }
+
+            AddText(MediaTypeHeaderValue.Parse(contentType));
+        }
+
+        /// <summary>
+        /// Adds a contentType to be used for logging as text.
+        /// </summary>
+        /// <param name="contentType">The content type to add.</param>
+        /// <param name="encoding">The encoding to use.</param>
+        public void AddText(string contentType, Encoding encoding)
+        {
+            if (contentType == null)
+            {
+                throw new ArgumentNullException(nameof(contentType));
+            }
+
+            if (encoding == null)
+            {
+                throw new ArgumentNullException(nameof(encoding));
+            }
+
+            var mediaType = MediaTypeHeaderValue.Parse(contentType);
+            mediaType.Encoding = encoding;
+            AddText(mediaType);
+        }
+
+        /// <summary>
+        /// Adds a <see cref="MediaTypeHeaderValue"/> to be used for logging as binary.
+        /// </summary>
+        /// <param name="mediaType">The MediaType to add.</param>
+        public void AddBinary(MediaTypeHeaderValue mediaType)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// Adds a content to be used for logging as text.
+        /// </summary>
+        /// <param name="contentType">The content type to add.</param>
+        public void AddBinary(string contentType)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// Clears all MediaTypes.
+        /// </summary>
+        public void Clear()
+        {
+            _mediaTypeStates.Clear();
+        }
+
+        internal class MediaTypeState
```

Comment:
Store the Encoding separately here, storing it in the MediaTypeHeaderValue is lossy since only the name is preserved, not the actual encoder.